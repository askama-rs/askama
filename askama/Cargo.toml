[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
