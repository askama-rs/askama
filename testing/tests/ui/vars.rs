// Check that `mut` can only be used with a variable name (not destructuring).

use askama::Template;

#[derive(Template)]
#[template(source = r#"
{% let mut (a, b) = (1, 2) %}
"#, ext = "html")]
struct Var1;

#[derive(Template)]
#[template(source = r#"
{% let mut [a, b] = [1, 2] %}
"#, ext = "html")]
struct Var2;

#[derive(Template)]
#[template(source = r#"
{% let mut Some(a) = Some("a") %}
"#, ext = "html")]
struct Var3;

fn main() {}
