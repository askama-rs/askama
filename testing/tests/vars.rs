#![allow(clippy::useless_let_if_seq)]

use askama::Template;

#[test]
fn test_let() {
    #[derive(Template)]
    #[template(source = "{% let v = s %}{{ v }}", ext = "txt")]
    struct LetTemplate<'a> {
        s: &'a str,
    }

    let t = LetTemplate { s: "foo" };
    assert_eq!(t.render().unwrap(), "foo");
}

#[test]
fn test_let_tuple() {
    #[derive(Template)]
    #[template(path = "let.html")]
    struct LetTupleTemplate<'a> {
        s: &'a str,
        t: (&'a str, &'a str),
    }

    let t = LetTupleTemplate {
        s: "foo",
        t: ("bar", "bazz"),
    };
    assert_eq!(t.render().unwrap(), "foo\nbarbazz");
}

#[test]
fn test_let_decl() {
    #[derive(Template)]
    #[template(path = "let-decl.html")]
    struct LetDeclTemplate<'a> {
        cond: bool,
        s: &'a str,
    }

    let t = LetDeclTemplate {
        cond: false,
        s: "bar",
    };
    assert_eq!(t.render().unwrap(), "bar");
}

#[test]
fn test_let_shadow() {
    #[derive(Template)]
    #[template(path = "let-shadow.html")]
    struct LetShadowTemplate {
        cond: bool,
    }

    impl LetShadowTemplate {
        fn tuple() -> (i32, i32) {
            (4, 5)
        }
    }

    let t = LetShadowTemplate { cond: true };
    assert_eq!(t.render().unwrap(), "22-1-33-11-22");

    let t = LetShadowTemplate { cond: false };
    assert_eq!(t.render().unwrap(), "222-1-333-4-5-11-222");
}

#[test]
fn test_self_iter() {
    #[derive(Template)]
    #[template(source = "{% for v in self.0 %}{{ v }}{% endfor %}", ext = "txt")]
    struct SelfIterTemplate(Vec<usize>);

    let t = SelfIterTemplate(vec![1, 2, 3]);
    assert_eq!(t.render().unwrap(), "123");
}

#[test]
fn test_if_let() {
    #[derive(Template)]
    #[template(
        source = "{% if true %}{% let t = a.unwrap() %}{{ t }}{% endif %}",
        ext = "txt"
    )]
    struct IfLet {
        a: Option<&'static str>,
    }

    let t = IfLet { a: Some("foo") };
    assert_eq!(t.render().unwrap(), "foo");
}

#[test]
fn test_destruct_tuple() {
    #[derive(Template)]
    #[template(path = "let-destruct-tuple.html")]
    struct LetDestructTupleTemplate {
        abcd: (char, ((char, char), char)),
    }

    let t = LetDestructTupleTemplate {
        abcd: ('w', (('x', 'y'), 'z')),
    };
    assert_eq!(t.render().unwrap(), "wxyz\nwz\nw");
}

#[test]
fn test_decl_range() {
    #[derive(Template)]
    #[template(
        source = "{% let x = 1 %}{% for x in x..=x %}{{ x }}{% endfor %}",
        ext = "txt"
    )]
    struct DeclRange;

    let t = DeclRange;
    assert_eq!(t.render().unwrap(), "1");
}

#[test]
fn test_decl_assign_range() {
    #[derive(Template)]
    #[template(
        source = "{% let x %}{% let x = 1 %}{% for x in x..=x %}{{ x }}{% endfor %}",
        ext = "txt"
    )]
    struct DeclAssignRange;

    let t = DeclAssignRange;
    assert_eq!(t.render().unwrap(), "1");
}

#[test]
fn test_not_moving_fields_in_var() {
    #[derive(Template)]
    #[template(
        source = "
{%- set t = title -%}
{{t}}/{{title -}}
",
        ext = "txt"
    )]
    struct DoNotMoveFields {
        title: String,
    }

    let x = DoNotMoveFields {
        title: "a".to_string(),
    };
    assert_eq!(x.render().unwrap(), "a/a");
}
