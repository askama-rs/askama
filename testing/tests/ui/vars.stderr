error: you can only use the `mut` keyword with a variable name
 --> <source attribute>:2:7
       "mut (a, b) = (1, 2) %}\n"
 --> tests/ui/vars.rs:6:21
  |
6 |   #[template(source = r#"
  |  _____________________^
7 | | {% let mut (a, b) = (1, 2) %}
8 | | "#, ext = "html")]
  | |__^

error: you can only use the `mut` keyword with a variable name
 --> <source attribute>:2:7
       "mut [a, b] = [1, 2] %}\n"
  --> tests/ui/vars.rs:12:21
   |
12 |   #[template(source = r#"
   |  _____________________^
13 | | {% let mut [a, b] = [1, 2] %}
14 | | "#, ext = "html")]
   | |__^

error: you can only use the `mut` keyword with a variable name
 --> <source attribute>:2:7
       "mut Some(a) = Some(\"a\") %}\n"
  --> tests/ui/vars.rs:18:21
   |
18 |   #[template(source = r#"
   |  _____________________^
19 | | {% let mut Some(a) = Some("a") %}
20 | | "#, ext = "html")]
   | |__^
